src/coreclr/src/vm/classlayoutinfo.cpp

```diff
@@ -0,0 +1,1108 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "class.h"
+#include "fieldmarshaler.h"
+
+#ifndef DACCESS_COMPILE
+
+namespace
+{
+    void SetOffsetsAndSortFields(
+        IMDInternalImport* pInternalImport,
+        const mdTypeDef cl,
+        LayoutRawFieldInfo* pFieldInfoArray,
+        const ULONG cInstanceFields,
+        const BOOL fExplicitOffsets,
+        const UINT32 cbAdjustedParentLayoutNativeSize,
+        Module* pModule,
+        LayoutRawFieldInfo** pSortArrayOut
+    )
+    {
+        HRESULT hr;
+        MD_CLASS_LAYOUT classlayout;
+        hr = pInternalImport->GetClassLayoutInit(cl, &classlayout);
+        if (FAILED(hr))
+        {
+            COMPlusThrowHR(hr, BFA_CANT_GET_CLASSLAYOUT);
+        }
+
+        LayoutRawFieldInfo* pfwalk = pFieldInfoArray;
+        mdFieldDef fd;
+        ULONG ulOffset;
+        while (SUCCEEDED(hr = pInternalImport->GetClassLayoutNext(
+            &classlayout,
+            &fd,
+            &ulOffset)) &&
+            fd != mdFieldDefNil)
+        {
+            // watch for the last entry: must be mdFieldDefNil
+            while ((mdFieldDefNil != pfwalk->m_MD) && (pfwalk->m_MD < fd))
+                pfwalk++;
+
+            // if we haven't found a matching token, it must be a static field with layout -- ignore it
+            if (pfwalk->m_MD != fd) continue;
+
+            if (!fExplicitOffsets)
+            {
+                // ulOffset is the sequence
+                pfwalk->m_sequence = ulOffset;
+            }
+            else
+            {
+                // ulOffset is the explicit offset
+                pfwalk->m_placement.m_offset = ulOffset;
+                pfwalk->m_sequence = (ULONG)-1;
+
+                // Treat base class as an initial member.
+                if (!SafeAddUINT32(&(pfwalk->m_placement.m_offset), cbAdjustedParentLayoutNativeSize))
+                    COMPlusThrowOM();
+            }
+        }
+        IfFailThrow(hr);
+
+        LayoutRawFieldInfo** pSortArrayEnd = pSortArrayOut;
+        // now sort the array
+        if (!fExplicitOffsets)
+        {
+            // sort sequential by ascending sequence
+            for (ULONG i = 0; i < cInstanceFields; i++)
+            {
+                LayoutRawFieldInfo** pSortWalk = pSortArrayEnd;
+                while (pSortWalk != pSortArrayOut)
+                {
+                    if (pFieldInfoArray[i].m_sequence >= (*(pSortWalk - 1))->m_sequence)
+                        break;
+
+                    pSortWalk--;
+                }
+
+                // pSortWalk now points to the target location for new LayoutRawFieldInfo*.
+                MoveMemory(pSortWalk + 1, pSortWalk, (pSortArrayEnd - pSortWalk) * sizeof(LayoutRawFieldInfo*));
+                *pSortWalk = &pFieldInfoArray[i];
+                pSortArrayEnd++;
```

Comment:
Any reason to use a open-coded sorting function here that looks like quadratic?

Reply:
Nope. That's just how it had been beforehand and I hadn't changed it when refactoring.